### PR TITLE
[mlir][TilingInterface] Allow controlling what fusion is done within tile and fuse

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Currently the `tileConsumerAndFuseProducerGreedilyUsingSCFFor` method greedily fuses through all slices that are generated during the tile and fuse flow. That is not the normal use case. Ideally the caller would like to control which slices get fused and which dont. This patch introduces a new field to the `SCFTileAndFuseOptions` to specify this control.

The contol function also allows the caller to specify if the replacement for the fused producer needs to be yielded from within the tiled computation. This allows replacing the fused producers in case they have other uses. Without this the original producers still survive negating the utility of the fusion.

The change here also means that the name of the function `tileConsumerAndFuseProducerGreedily...` can be updated. Defering that to a later stage to reduce the churn of API changes.

### CODE REVIEW

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h [hunk 1]
```diff
@@ -97,6 +97,30 @@ struct SCFTileAndFuseOptions {
...
+  /// Control function to check if a slice needs to be fused or not,
+  /// The control function recieves
+  /// 1) the slice along which fusion is to be done,
+  /// 2) the producer value that is to be fused
+  /// 3) a boolean value set to `true` if the fusion is from
+  ///    a destination operand.
+  /// It retuns two booleans
+  /// - returns `true` if the fusion should be done through the candidate slice
+  /// - returns `true` if a replacement for the fused producer needs to be
+  ///   yielded from within the tiled loop. Note that is is valid to return
```
Comment (hanhanW) [inline mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:110]: nit: typo

```suggestion
  ///   yielded from within the tiled loop. Note that it is valid to return
```

**Review by hanhanW — APPROVED** [review]
The code looks okay to me. I was thinking if all the cases are covered or not, then I found that the answer is yes: https://github.com/llvm/llvm-project/blob/main/mlir/test/Interfaces/TilingInterface/tile-and-fuse-using-interface.mlir

This looks cool to me because it internalize some logics through control function. It is easier to use.

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h [hunk 2]
```diff
@@ -97,6 +97,30 @@ struct SCFTileAndFuseOptions {
     tilingOptions = options;
     return *this;
   }
+
+  /// Control function to check if a slice needs to be fused or not,
+  /// The control function recieves
```
Comment (chelini) [inline mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:102]: nit: receives

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h [hunk 3]
```diff
@@ -97,6 +97,30 @@ struct SCFTileAndFuseOptions {
...
+  /// 3) a boolean value set to `true` if the fusion is from
+  ///    a destination operand.
+  /// It retuns two booleans
+  /// - returns `true` if the fusion should be done through the candidate slice
+  /// - returns `true` if a replacement for the fused producer needs to be
+  ///   yielded from within the tiled loop. Note that is is valid to return
+  ///   `true` only if the slice fused is disjoint across all iterations of the
+  ///   tiled loop. It is upto the caller to ensure that this is true for the
+  ///   fused producers.
+  using ControlFnTy = std::function<std::tuple<bool, bool>(
```
Comment (chelini) [inline mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:114]: why not using std::pair?

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h [hunk 4]
```diff
@@ -97,6 +97,30 @@ struct SCFTileAndFuseOptions {
...
+  /// 1) the slice along which fusion is to be done,
+  /// 2) the producer value that is to be fused
+  /// 3) a boolean value set to `true` if the fusion is from
+  ///    a destination operand.
+  /// It retuns two booleans
+  /// - returns `true` if the fusion should be done through the candidate slice
+  /// - returns `true` if a replacement for the fused producer needs to be
+  ///   yielded from within the tiled loop. Note that is is valid to return
+  ///   `true` only if the slice fused is disjoint across all iterations of the
+  ///   tiled loop. It is upto the caller to ensure that this is true for the
```
Comment (chelini) [inline mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:112]: nit: up to

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h [hunk 5]
```diff
@@ -97,6 +97,30 @@ struct SCFTileAndFuseOptions {
     tilingOptions = options;
     return *this;
   }
+
+  /// Control function to check if a slice needs to be fused or not,
+  /// The control function recieves
+  /// 1) the slice along which fusion is to be done,
+  /// 2) the producer value that is to be fused
+  /// 3) a boolean value set to `true` if the fusion is from
```
Comment (chelini) [inline mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:105]: Why is this necessary? Can you please provide more details on the use case you are considering?

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h [hunk 6]
```diff
@@ -97,6 +97,30 @@ struct SCFTileAndFuseOptions {
...
+  ///    a destination operand.
+  /// It retuns two booleans
+  /// - returns `true` if the fusion should be done through the candidate slice
+  /// - returns `true` if a replacement for the fused producer needs to be
```
Comment (chelini) [inline mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:109]: Same here. Why do you want to have this information? What is the implication of always returning false if we don't provide a callback (i.e., using fusion through transform?)?

**Comment by chelini** [comment]
> The code looks okay to me. I was thinking if all the cases are covered or not, then I found that the answer is yes: https://github.com/llvm/llvm-project/blob/main/mlir/test/Interfaces/TilingInterface/tile-and-fuse-using-interface.mlir
> 
> This looks cool to me because it internalize some logics through control function. It is easier to use.

Indeed, thanks for working on this @MaheshRavishankar.  A callback to control what to fuse is good, in my opinion.

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h (hunk 5 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:105]: I dont have any right now, but trying to be future proof. Its seems useful to have the information that the fusion is happening for a destination operand. You can of course retrieve this information from the uses of `candidateSlice`

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h (hunk 6 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:109]: If it is `false` (which is the default) the producers just get fused into the tile loop nest. If you look at https://github.com/llvm/llvm-project/blob/main/mlir/test/Interfaces/TilingInterface/tile-fuse-and-yield-using-interface.mlir you will see an example where it is useful for this to be `true`.

File: mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h (hunk 3 above)
Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/SCF/Transforms/TileUsingInterface.h:114]: no specific reason. Tuple seems like a more "newer" approach. No difference really.

**Review by chelini — APPROVED** [review]
Ok thanks for clarifying this. The PR looks good to me.
